### PR TITLE
MM-375 Adds markdown support using marked js library

### PR DESCRIPTION
**Pull request by rgarmsen2295**
## _Markdown Support_

This pull request adds markdown support to anywhere that the textToJsx utility function is called (namely posts and the channel description). Also adds help text under the textbox in the center pane and RHS after a user has typed at least one word informing them of some simple markdown commands with a link for more advanced ones (currently pointing to a GitHub markdown help page, will be changed later).  Primary weirdness (like last edit timestamps) are due to needing to unmount and remount react components when the html is changed due to React not liking the DOM being changed without this.  By modifying the component's key the component is forced to do this.

As far as what's quirky and different from plain 'ole markdown

* No markdown links, e.g. using "[Example Link](http://example.net/)" to format a link, we have another library for parsing links that we just got working properly, didn't want to rock the boat too much (some users also complained about the implementation from what I read, but still would be good to have, see below)
* No markdown inline images, similar to above it didn't work and we already have our own file upload system
* No mangling of emails, doubt it would've seen much use (only even tested it towards the end), but it basically wanted to modify the React DOM every time you did anything causing React to panic and crash
* No tables, not part of the standard markdown syntax and were not working
* No directly entering html, we sanitize that, not sanitizing that would be bad

...otherwise it's pretty much what you would expect, when I had the chance I went for some more modern options such as using "GitHub" style paragraphs/line breaks (just do a newline instead of leaving two empty spaces at the end of a line).  

Next steps would be getting markdown links and possibly tables working, I think the links feature is really cool and useful and probably wouldn't be too much work to get going.  Tables would probably be more work and is not standard but is something to think about.  Also edited the config file on this one.

### CODE REVIEW

File: web/react/utils/utils.jsx [hunk 1]
```diff
@@ -95,15 +96,20 @@ module.exports.getCookie = function(name) {
   if (parts.length == 2) return parts.pop().split(";").shift();
 }
 
-
 module.exports.notifyMe = function(title, body, channel) {
-  if ("Notification" in window && Notification.permission !== 'denied') {
+    if ("Notification" in window && Notification.permission !== 'denied') {
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: Obviously we're going to want to reindent this file as part of a later pull request, but this line is still off since it's indented the same as the following line which is inside the if block.

File: web/react/components/channel_header.jsx [hunk 2]
```diff
@@ -95,6 +95,7 @@ module.exports = React.createClass({
     _onChange: function() {
         var newState = getStateFromStores();
         if (!utils.areStatesEqual(newState, this.state)) {
+            newState.last_edit_time = new Date();
```
Comment (hmhealey) [inline web/react/components/channel_header.jsx]: We might want to use Date.now instead of new Date objects for keeping track of the last edit time since it's faster.

File: web/react/utils/utils.jsx [hunk 3]
```diff
@@ -385,11 +391,76 @@ module.exports.searchForTerm = function(term) {
...
+var endTagRegex = /(<\s*\/\s*\w\s*.*?>|<\s*br\s*>)+/g;
 
 module.exports.textToJsx = function(text, options) {
+    var useMarkdown = config.AllowMarkdown;
+
+    if (useMarkdown) {
+        text = marked(text, {sanitize: true, mangle: false, gfm: true, breaks: true, tables: false, smartypants: true, renderer: module.exports.customMarkedRenderer()});
+    }
+
+    console.log(text);
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: There's a leftover debug statement here.

File: web/react/utils/utils.jsx [hunk 4]
```diff
@@ -437,60 +528,86 @@ module.exports.textToJsx = function(text, options) {
...
                     highlightSearchClass = " search-highlight";
                 }
 
-                inner.push(<span key={name+i+z+"_span"}>{prefix}<a className={mClass + highlightSearchClass + " mention-link"} key={name+i+z+"_link"} href="#" onClick={function(value) { return function() { module.exports.searchForTerm(value); } }(name)}>@{name}</a>{suffix} </span>);
+                if (useMarkdown) {
+                    prefixSpan ? inner.push(prefixSpan) : null;
+                    inner.push(<span key={word+i+z+"word_span"}><a className={mClass + highlightSearchClass + " mention-link"} key={name+i+z+"_link"} href="#" dangerouslySetInnerHTML={{__html: "@" + name}} onClick={function(value) { return function() { module.exports.searchForTerm(value); } }(name)} /></span>);
+                    suffixSpan ? inner.push(suffixSpan) : null;
+                }
+                else
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: Not sure when we'll be fully enforcing the style guide, but this else and some of the later ones should have the curly braces fixed so that it matches when we do switch it on.

File: web/react/utils/utils.jsx [hunk 5]
```diff
@@ -437,60 +528,86 @@ module.exports.textToJsx = function(text, options) {
...
+                    inner.push(<span key={name+i+z+"_span"}>{prefix}<a className={mClass + highlightSearchClass + " mention-link"} key={name+i+z+"_link"} href="#" onClick={function(value) { return function() { module.exports.searchForTerm(value); } }(name)}>@{name}</a>{suffix} </span>);
             } else if (testUrlMatch(word).length) {
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: Does marked have a URL parser that we could use instead of our version? Unless there's something that ours does differently, it might make sense to just switch to using theirs.

File: web/react/utils/utils.jsx [hunk 6]
```diff
@@ -420,12 +491,32 @@ module.exports.textToJsx = function(text, options) {
...
-            var trimWord = word.replace(puncStartRegex, '').replace(puncEndRegex, '').trim();
+            var trimWord;
+            if (useMarkdown) {
+                trimWord = word.replace(endTagRegex, '').replace(startTagRegex, '').replace(puncStartRegex, '').replace(puncEndRegex, '').trim();
+            } else {
+                trimWord = word.replace(puncStartRegex, '').replace(puncEndRegex, '').trim();
+            }
             var mentionRegex = /^(?:@)([a-z0-9_]+)$/gi; // looks loop invariant but a weird JS bug needs it to be redefined here
             var explicitMention = mentionRegex.exec(trimWord);
+            console.log(word);
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: Another debug statement

File: web/react/utils/utils.jsx [hunk 7]
```diff
@@ -420,12 +491,32 @@ module.exports.textToJsx = function(text, options) {
...
+            var prefix;
+            var suffix;
+            var prefixSpan;
+            var suffixSpan;
+            if (useMarkdown) {
+                prefix = (word.match(startTagRegex) ? word.match(startTagRegex) : "") + (word.replace(startTagRegex, '').match(puncStartRegex) ? word.replace(startTagRegex, '').match(puncStartRegex) : "");
+                suffix = (word.match(endTagRegex) ? word.match(endTagRegex) : "") + (word.replace(endTagRegex, '').match(puncEndRegex) ? word.replace(endTagRegex, '').match(puncEndRegex) : "");
+                prefixSpan = prefix ? (<span key={word+i+z+"pre_span"}><span dangerouslySetInnerHTML={{__html: prefix}} /></span>) : null;
+                suffixSpan = suffix ? (<span key={word+i+z+"suf_span"}><span dangerouslySetInnerHTML={{__html: suffix}} /> </span>) : (<span key={word+i+z+"suf_span"}> </span>);
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: I'm having difficulty understanding this part (which may mean that it should be changed), but if I'm understanding it correctly, if marked outputs something like "<b>...</b>", this will turn it into something like "<span><span><b></span></span>...<span><span></b></span></span>" which doesn't seem like it should work.

File: web/react/utils/utils.jsx [hunk 8]
```diff
@@ -437,60 +528,86 @@ module.exports.textToJsx = function(text, options) {
...
-                var prefix = word.match(puncStartRegex);
-
                 if (searchTerm === name) {
                     highlightSearchClass = " search-highlight";
                 }
 
-                inner.push(<span key={name+i+z+"_span"}>{prefix}<a className={mClass + highlightSearchClass + " mention-link"} key={name+i+z+"_link"} href="#" onClick={function(value) { return function() { module.exports.searchForTerm(value); } }(name)}>@{name}</a>{suffix} </span>);
+                if (useMarkdown) {
+                    prefixSpan ? inner.push(prefixSpan) : null;
+                    inner.push(<span key={word+i+z+"word_span"}><a className={mClass + highlightSearchClass + " mention-link"} key={name+i+z+"_link"} href="#" dangerouslySetInnerHTML={{__html: "@" + name}} onClick={function(value) { return function() { module.exports.searchForTerm(value); } }(name)} /></span>);
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: Why is this part using dangerouslySetInnerHTML? It could just nest the text like it does below.

File: web/react/utils/utils.jsx [hunk 9]
```diff
@@ -437,60 +528,86 @@ module.exports.textToJsx = function(text, options) {
...
-                var suffix = word.substring(word.indexOf(match.text)+match.text.length);
-
-                inner.push(<span key={word+i+z+"_span"}>{prefix}<a key={word+i+z+"_link"} className={"theme" + highlightSearchClass} target="_blank" href={link}>{match.text}</a>{suffix} </span>);
+                prefix = word.substring(0,word.indexOf(match.text));
+                suffix = word.substring(word.indexOf(match.text)+match.text.length);
+                prefixSpan = prefix ? (<span key={word+i+z+"pre_span"}><span dangerouslySetInnerHTML={{__html: prefix}} /></span>) : null;
+                suffixSpan = suffix ? (<span key={word+i+z+"suf_span"}><span dangerouslySetInnerHTML={{__html: suffix}} /> </span>) : inner.push(<span key={word+i+z+"suf_span"}> </span>);
 
+                if (useMarkdown) {
+                    prefixSpan ? inner.push(prefixSpan) : null;
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: I think this should be changed to a normal if statement. I don't remember our stance on ternary statements, but using them as shorthand ifs without using their return value seems very odd to me.

File: web/react/utils/utils.jsx [hunk 10]
```diff
@@ -437,60 +528,86 @@ module.exports.textToJsx = function(text, options) {
...
+                    else
+                        inner.push(<span key={word+i+z+"_span"}>{prefix}<span className={mentionClass + highlightSearchClass}>{module.exports.replaceHtmlEntities(trimWord)}</span>{suffix} </span>);
                 }
 
             } else if (word === "") {
                 // if word is empty dont include a span
             } else {
-                inner.push(<span key={word+i+z+"_span"}><span className={highlightSearchClass}>{module.exports.replaceHtmlEntities(word)}</span> </span>);
+                if (useMarkdown)
+                    inner.push(<span key={word+i+z+"_span"}><span className={highlightSearchClass} dangerouslySetInnerHTML={{__html: word}} /> </span>);
```
Comment (hmhealey) [inline web/react/utils/utils.jsx]: Like my comment on line 511, this could lead to weird nesting of tags if marked were to emit something where there is any whitespace between the start and end tags.

File: web/react/utils/utils.jsx (hunk 5 above)
Comment (hmhealey) [inline web/react/utils/utils.jsx]: Oh sorry, I just noticed your comment on the pull request about the URL parsing. Never mind on this one.

**Comment by hmhealey** [comment]
I'm starting to think we may want to rewrite the whole textToJsx function with something more suitable to applying complex formatting (like using a state machine). The function is already rather big and complex and once markdown is in, it'll be even bigger. I was already coming across issues with the stateless way we're handling parsing when I was trying to fix MM-1706.

**Comment by hmhealey** [comment]
Also, what is the stuff tracking the last time a post was edited used for? I can't find anything it's used for in the markdown handling code.

File: web/react/utils/utils.jsx (hunk 1 above)
Comment (rgarmsen2295) [inline web/react/utils/utils.jsx]: Yeah for this and your other comments on the style guide stuff I left a note in the dev room in Cayley that I think may have gotten pushed up, but I was still planning on doing at least one more final pass through for style guide stuff (and removing stuff like the debug statements, etc).

File: web/react/components/channel_header.jsx (hunk 2 above)
Comment (rgarmsen2295) [inline web/react/components/channel_header.jsx]: Huh, had no idea, ok yeah I'll take a look at using that.

File: web/react/utils/utils.jsx (hunk 7 above)
Comment (rgarmsen2295) [inline web/react/utils/utils.jsx]: I didn't think about it like that, but it does in fact work!  Basically what it does is rips out the html tags from the start and/or the end of whatever the current word is and puts them in their own separate span.  Why?  Well whenever you try to do things like **@alice** or **##testing** things tended to start breaking due to how we handle making them links and such.  The easiest way I found to consistently make it work was to just remove the markdown tags from the equation while parsing these specific cases, using the prefix/suffix outline that we already had going.  The long replace/match string that also uses multiple ternaries is pretty hard to parse I admit, I need to refactor it to use if else statements for the style guide anyway so hopefully that will make it a bit clearer.  If it's still too confusing at that point I'll try to make it clearer somehow.

File: web/react/utils/utils.jsx (hunk 8 above)
Comment (rgarmsen2295) [inline web/react/utils/utils.jsx]: Pretty sure you're right, since I took out the actual html tags from the word in these cases they should no longer need dangerouslySetHTML, I'll try that out thanks!

**Comment by rgarmsen2295** [comment]
I have to admit I don't have a lot of experience writing large parsers such as this, but you may be right (had a heck of a time getting some of this to work properly) as it is getting confusingly large and unwieldy.  Not quite sure what next steps would be but definitely a good thing to start talking about

For the post edit stuff I mentioned it in the first paragraph of my description at the top but basically React hates it when you try to edit certain DOM elements, causing it to crash and panic (e.g. when editing a post with a blockquote), unless you unmount the component and remount it before rendering it again.  Changing the component's key by a constantly increasing edit time forces this to occur (took Joram and I a long time to figure that out).

**Comment by hmhealey** [comment]
That's good to know about React's weird behaviour when editing elements since I might have to worry about that when I re-add and improve ^replying.

And yeah for the parsing, I don't have a ton of experience with parsing either so we can discuss it in a bit more detail later this week. I know we have a few layers of parsing that we're currently doing, but there must be some cleaner way of doing this or at least combining them. Hopefully without going to the extent of having to build a full-on parser with a tokenizer and stuff.

**Comment by rgarmsen2295** [comment]
Yeah, really bizarre behavior, hopefully it's an issue they resolve at some point.  And agreed, long-term it's going to probably become problematic if we try to bolt too much more onto it, but yeah not sure how to make the cost of changing it to be lower than a full on expensive rebuild.

**Comment by crspeller** [comment]
@rgarmsen2295 What's the status of this?

**Comment by rgarmsen2295** [comment]
@crspeller Would still like to get a bit of pre-testing in via a dev instance, but regardless of whether or not that happens, from talking to @hmhealey this is pretty much ready to go as is with the idea of me going back and style guide refactoring it on a later pass since it already changes so much (the later pass most likely also adding in markdown 'link functionality at the same time).  So assuming two give their official +1's I would say merge it in today or tomorrow (original plan with @coreyhulen was to get it in early in the release cycle for as much testing as possible on CI).  Things will break but better to find out what does early and see how much stress this adds performance wise.

**Comment by rgarmsen2295** [comment]
Ok @crspeller and @hmhealey made additional changes according to issues found in test environment, please take another look (specifically the last two commits) at the changes when possible, thanks!

**Comment by hmhealey** [comment]
Looks good to me

**Comment by rgarmsen2295** [comment]
Closing due to changing of spec from Markdown support to a simplified version supporting **bold**, _italics_, and `code`
